golem-worker-executor-base/src/services/rdbms/sqlx_common.rs

```diff
@@ -0,0 +1,867 @@
+                            e
+                        );
+                        e
+                    })?;
+                    Ok(Arc::new(result))
+                })
+            })
+            .await?;
+
+        self.pool_workers_cache
+            .entry(key.clone())
+            .or_default()
+            .insert(worker_id.clone());
+
+        Ok(pool)
+    }
+
+    #[allow(dead_code)]
+    pub(crate) async fn remove_pool(&self, key: &RdbmsPoolKey) -> Result<bool, Error> {
+        let _ = self.pool_workers_cache.remove(key);
+        let pool = self.pool_cache.try_get(key);
+        if let Some(pool) = pool {
+            self.pool_cache.remove(key);
+            pool.close().await;
+            Ok(true)
+        } else {
+            Ok(false)
+        }
+    }
+
+    fn record_metrics<R>(
+        &self,
+        name: &'static str,
+        start: Instant,
+        result: Result<R, Error>,
+    ) -> Result<R, Error> {
+        record_rdbms_metrics(&self.rdbms_type, name, start, result)
+    }
+}
+
+#[async_trait]
+impl<T, DB> Rdbms<T> for SqlxRdbms<T, DB>
+where
+    T: RdbmsType + Sync + QueryExecutor<T, DB> + 'static,
+    DB: Database,
+    for<'c> &'c mut <DB as Database>::Connection: sqlx::Executor<'c, Database = DB>,
+    RdbmsPoolKey: PoolCreator<DB>,
+{
+    async fn create(&self, address: &str, worker_id: &WorkerId) -> Result<RdbmsPoolKey, Error> {
+        let start = Instant::now();
+
+        let result = {
+            let key = RdbmsPoolKey::from(address).map_err(Error::ConnectionFailure)?;
+            info!(
+                rdbms_type = self.rdbms_type.to_string(),
+                pool_key = key.to_string(),
+                "create connection",
+            );
+
+            let _ = self.get_or_create(&key, worker_id).await?;
+
+            Ok(key)
+        };
+        self.record_metrics("create-connection", start, result)
+    }
+
+    fn remove(&self, key: &RdbmsPoolKey, worker_id: &WorkerId) -> bool {
+        match self.pool_workers_cache.get_mut(key) {
+            Some(mut workers) => (*workers).remove(worker_id),
+            None => false,
+        }
+    }
+
+    fn exists(&self, key: &RdbmsPoolKey, worker_id: &WorkerId) -> bool {
+        self.pool_workers_cache
+            .get(key)
+            .is_some_and(|workers| workers.contains(worker_id))
+    }
+
+    async fn execute(
+        &self,
+        key: &RdbmsPoolKey,
+        worker_id: &WorkerId,
+        statement: &str,
+        params: Vec<T::DbValue>,
+    ) -> Result<u64, Error>
+    where
+        <T as RdbmsType>::DbValue: 'async_trait,
+    {
+        let start = Instant::now();
+        debug!(
+            rdbms_type = self.rdbms_type.to_string(),
+            pool_key = key.to_string(),
+            "execute - statement: {}, params count: {}",
+            statement,
+            params.len()
+        );
+
+        let result = {
+            let pool = self.get_or_create(key, worker_id).await?;
```

Comment:
I understand the db resource is created per worker instance. But in a typical golem architecture, a pool managed by each worker can lead to a situations that is impossible to handle. What do you think? Easily a 10000 workers, and even with 10 connections, it is 100000 connections. Where exactly are these configurations handled?  I understand pool size can be configured, but not quite convinced of the production gradeness of this architecture. If you could explain, that will be great.

A shared durable proxy for managing pools can be a typical architecture in larger systems, and golem can handle these orchestrations internally without making it a concern for the users.